### PR TITLE
Integration between eBPF and Apps

### LINKED ISSUE #8912: Copy functions from apps.plugin to ebpf.plugin
##### Feature idea summary
We have defined at #7873 that we will use a common shared library between both collectors, but my initial suggestion is we copy the necessary functions to `ebpf.plugin` and after this to do the shared library. This copy must be done to the main `thread` to be available for all eBPF threads that will be created on the collector. The following functions need to be copied:

- `am_i_running_as_root()` : This is a function that should be common for at least 7 collectors that netdata has, probably the correct place to it is inside `libnetdata`.
- `read_apps_groups_conf()` - The file `apps_group.conf` will be a common file for both collectors.
- `get_apps_groups_target()`:  find or create a new target, this function is called from  `read_apps_groups_conf()`.
- `zero_all_targets()`: Reset all targets with value zero.
- `aggregate_pid_on_target()` - This function does not need to be moved to the library, but its logic needs to be moved inside the collector.
- `post_aggregate_targets();` - Like `aggregate_pid_on_target()` we also need to move the logic to the collector.
- `get_users_target()` and `get_groups_target()`: Aggregate all processes in groups, to have a limited number of dimensions. 

We also will need to do some changes for the plugin to be able finding and to open `apps_group.conf`.

Some functions do call for specific `avl` trees and local variables that will also needed to be ported to eBPF plugin. When we move this functions, it is also good move the cited variables as arguments instead to keep them inside the functions.

##### Expected behavior
eBPF plugin must be able to parse `apps_groups.conf`.

### LINKED ISSUE #8914: Change threads to create charts inside `Apps` submenu
##### Feature idea summary
The eBPF programs are reading only hash table with global information from the thread, but they are also storing a PID statistics in another hash table. Netdata needs to read these information from kernel ring to make the charts.  To move this forward we need to do the following changes on the eBPF.plugin threads:

- Map the functions `bpf_map_get_next_key`, `bpf_map_delete_elem` and `bpf_map_lookup_elem` to get all PIDs  from the shared libraries to eBPF modules.
-  A function to create the charts when the collector is loaded, this chart will have the apps groups as dimensions.
- Another function that will read all the data from kernel ring to user ring. It will be interesting to use the 3 new functions mapped instead to use the current values present inside `/proc` to be sure we are reading all the information. 
- Finally we will have another function that will send new values for each dimension of the new charts.

The last two functions will be called inside the main loop of the thread.

##### Expected behavior
To have new charts inside apps.plugin.

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
Fixes  #8912
Fixes #8914

This PR brings the integration between the collectors eBPF and Apps.

##### Component Name
Collector
##### Test Plan
1 - Install Netdata using     `./netdata-installer.sh --dont-start-it`.
2 - Download the file [files4.zip](https://github.com/netdata/netdata/files/4762377/files4.zip) and extract it inside `/usr/libexec/netdata/plugins.d`.
3 - Run netdata and verify the charts on `apps.plugin`.
4 - Stop netdata after few seconds.
5 - Start netdata again.

Do the steps 4 and 5 at least 3 times. We should have consistency on all charts (general and apps) in all tests.

It is expected to see 14 charts inside apps submenu.

##### Additional Information

### CODE REVIEW

**Comment by ilyam8** [comment]
Neither `Summary` and `Test plan` are not clear. Part? What part?

> so we expect that the reviewer will compile and verify any possible erro

What kind of errors should we expect? It depends on changes? (go to _summary is not clear_).

**Comment by thiagoftsm** [comment]
> Neither `Summary` and `Test plan` are not clear. Part? What part?
> 
> > so we expect that the reviewer will compile and verify any possible erro
> 
> What kind of errors should we expect? It depends on changes? (go to _summary is not clear_).

This is the motive it is `[WIP]` we are finalizing it before to give details ;)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

File: collectors/ebpf.plugin/ebpf_apps.c [hunk 1]
```diff
@@ -956,19 +885,34 @@ static void cleanup_exited_pids(void) {
...
+                                   pid_t *index,
+                                   int (*bpf_map_get_next_key)(int, const void *, void *),
+                                   int (*bpf_map_lookup_elem)(int, const void *, void *),
+                                   int tbl_pid_stats_fd)
+#else
+int collect_data_for_all_processes(ebpf_process_stat_t **out,
+                                   pid_t *index,
+                                   int tbl_pid_stats_fd)
+#endif
 {
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_apps.c:983]: Why are you doing it this way? Why don't you assign function addresses in `process_functions` depending on `STATIC` macro?

Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_apps.c:983]: The `kickstart_static64` will need a different structure to compile, the `STATIC` is already used on `netdata/kernel-collector` when we need to drop completely the shared libraries, I am keeping this here for while to keep the same pattern used on the other repository.

Comment (vlvkobal) [reply collectors/ebpf.plugin/ebpf_apps.c:983]: I'm not against using `STATIC`. I just prefer to do it in one place rather than several ones.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

**Comment by thiagoftsm** [comment]
@vlvkobal I observed that we are having a small problem with the conditional variables. I inserted debug messages inside the function `ebpf_socket_send_apps_data()` and good part of the time that I was testing, the variables `root_pid` was cleaned, but this variable is used and set on `process` threads and the charts are created.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5589756&bid=18315734)

File: collectors/ebpf.plugin/ebpf_apps.c [hunk 2]
```diff
@@ -0,0 +1,1238 @@
...
+
+        for(p = root_of_pids; p ; p = p->next) {
+            if(unlikely(!p->sortlist && !p->children_count))
+                p->sortlist = sortlist++;
+
+            if(unlikely(
+                    !p->children_count            // if this process does not have any children
+                    && !p->merged                 // and is not already merged
+                    && p->parent                  // and has a parent
+                    && p->parent->children_count  // and its parent has children
```
Comment (underhood) [inline collectors/ebpf.plugin/ebpf_apps.c:815]: is this not redundant? If I have parent can my parent have 0 children?

File: collectors/ebpf.plugin/ebpf_apps.c [hunk 3]
```diff
@@ -0,0 +1,1238 @@
...
+    // link all children to their parents
+    // and update children count on parents
+    for(p = root_of_pids; p ; p = p->next) {
+        // for each process found
+
+        p->sortlist = 0;
+        p->parent = NULL;
+
+        if(unlikely(!p->ppid)) {
+            p->parent = NULL;
```
Comment (underhood) [inline collectors/ebpf.plugin/ebpf_apps.c:753]: you just set it to NULL 3 lines above?

File: collectors/ebpf.plugin/ebpf_apps.c (hunk 2 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_apps.c:815]: Probably this comment is not appropriate when read only the first part of the sentence, but the group looks like more acceptable:

```
// and its parent has children
// and the target of this process and its parent is the same,
// or the parent does not have a target
```

This was completely copied from `apps_plugin.c` without any change.

File: collectors/ebpf.plugin/ebpf_apps.c (hunk 3 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_apps.c:753]: Good catch, it really looks like an unnecessary code from `apps plugin.c`.

File: collectors/ebpf.plugin/ebpf.c [hunk 4]
```diff
@@ -533,6 +535,36 @@ void ebpf_print_help() {
...
+/**
+ * Start Ptherad Variable
+ *
+ * This function starts all pthread variables.
+ *
+ * @return It returns 0 on success and -1.
+ */
+int ebpf_start_pthread_variables()
+{
+    if (pthread_mutex_init(&lock, NULL)) {
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf.c:549]: BTW
> pthread_mutex_init always returns 0

https://manned.org/pthread_mutex_init.3#head6

**Comment by vlvkobal** [comment]
The plugin shows inconsistent data.
A lot of charts/dimensions are always zero:
- `ebpf.file_descriptor->open`
- `ebpf.file_descriptor->thread`
- `ebpf.file_descriptor`
- `ebpf.file_descriptor` (sometimes I can see it working)
- `apps.vfs_write_bytes`
- `apps.vfs_read_bytes`
- `apps.vfs_read_bytes`
- `apps.task_close`

**Comment by vlvkobal** [comment]
There are error messages every second:
```
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_open' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_closed' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_deleted' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_write_call' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_read_call' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_write_bytes' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_read_bytes' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.process_create' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.thread_create' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.task_close' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.bandwidth_sent' on host 'arch-esxi' already exists.
2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.bandwidth_recv' on host 'arch-esxi' already exists.
```

**Comment by vlvkobal** [comment]
Every time on Netdata restart `/sys/kernel/debug/tracing/kprobe_events` should be cleaned in order to have working eBPF charts.

File: collectors/ebpf.plugin/ebpf.c (hunk 4 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf.c]: I removed them, this is so common that I wrote the check in the automatic.

**Comment by thiagoftsm** [comment]
> The plugin shows inconsistent data.
> A lot of charts/dimensions are always zero:
> 
>     * `ebpf.file_descriptor->open`
>     * `ebpf.file_descriptor->thread`
>     * `ebpf.file_descriptor`
>     * `ebpf.file_descriptor` (sometimes I can see it working)
>     * `apps.vfs_write_bytes`
>     * `apps.vfs_read_bytes`
>     * `apps.vfs_read_bytes`
>     * `apps.task_close`

I saw this, and I created a PR to restart the collector when it is not collecting enough data.
I read yesterday in some [comments](https://github.com/iovisor/bcc/issues/1016#issuecomment-285132782)  on other project that called my attention. This behavior is expected and we will need to move for tracepoints when it is possible.

**Comment by thiagoftsm** [comment]
> There are error messages every second:
> 
> ```
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_open' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_closed' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.file_deleted' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_write_call' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_read_call' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_write_bytes' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.vfs_read_bytes' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.process_create' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.thread_create' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.task_close' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.bandwidth_sent' on host 'arch-esxi' already exists.
> 2020-06-09 13:49:01: netdata INFO  : PLUGINSD[ebpf] : RRDSET: chart name 'apps.bandwidth_recv' on host 'arch-esxi' already exists.
> ```

I had commented a line to test and I forgot to return it, I fixed this in the last commit.

**Comment by vlvkobal** [comment]
> I saw this, and I created a PR to restart the collector when it is not collecting enough data.

It doesn't help.

**Comment by thiagoftsm** [comment]
> > I saw this, and I created a PR to restart the collector when it is not collecting enough data.
> 
> It doesn't help.

What kernel are you using now? Do you have the messages "ebpf.plugin is losing many probes, plugin will be restarted." inside your `error.log`?

**Comment by vlvkobal** [comment]
Kernel 5.6.15
> Do you have the messages "ebpf.plugin is losing many probes

```
2020-06-09 15:25:53: ebpf.plugin ERROR : EBPF PROCESS : ebpf.plugin is losing many probes, plugin will be restarted. (errno 2, No such file or directory)
2020-06-09 15:25:53: ebpf.plugin INFO  : EBPF SOCKET : thread with task id 170123 finished
2020-06-09 15:25:53: ebpf.plugin INFO  : EBPF PROCESS : thread with task id 170122 finished
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF PROCESS : thread created with task id 170353
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF SOCKET : thread created with task id 170354
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF SOCKET : set name of thread 170354 to EBPF SOCKET
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF PROCESS : set name of thread 170353 to EBPF PROCESS
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF SOCKET : Cannot load library /usr/libexec/netdata/plugins.d/libnetdata_ebpf.so.5.6.15-arch1-1
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF SOCKET : Default shared library /usr/libexec/netdata/plugins.d/libnetdata_ebpf.so.5.4.20 loaded with success.
[eBPF] failed to create kprobe 'tcp_sendmsg' error 'File exists'
2020-06-09 15:25:55: ebpf.plugin ERROR : EBPF SOCKET : Cannot load program: /usr/libexec/netdata/plugins.d/pnetdata_ebpf_socket.5.4.20.o (errno 17, File exists)
2020-06-09 15:25:55: ebpf.plugin INFO  : EBPF SOCKET : thread with task id 170354 finished
```

**Comment by vlvkobal** [comment]
image

**Comment by thiagoftsm** [comment]
> image

This problem was fixed, at least on kernel 5.4.45, I intend to test in other kernels today. The function `bpf_map_get_next_key` was creating all problem for us.

File: collectors/ebpf.plugin/ebpf.c [hunk 5]
```diff
@@ -302,19 +302,23 @@ void write_io_chart(char *chart, char *family, char *dwrite, char *dread, netdat
...
  */
 void ebpf_write_chart_cmd(char *type
     , char *id
+    , char *title
     , char *units
     , char *family
     , int order)
 {
-    printf("CHART %s.%s '' 'title' '%s' '%s' '' line %d %d\n",
+    (void)title;
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf.c:317]: Did you forget to delete it?

**Comment by underhood** [comment]
`apps.task_close` is still stuck at 0 in CentOS 8 but it works on ubuntu.

Centos:
Screenshot from 2020-06-11 10-55-00

on bubuntu it looks more healthy:
Screenshot from 2020-06-11 10-56-21

File: collectors/ebpf.plugin/ebpf.c (hunk 5 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf.c]: You are right, I missed to remove, thank you!

**Comment by thiagoftsm** [comment]
> Why all the charts still have no titles? I keep asking it for months smile

I finally brought them, thank you to call my attention for this.

**Comment by thiagoftsm** [comment]
>     * Why did you make it so?
Because, this was the first collector I wrote for Netdata. ;) 

>     * Why `update_every` is hardcoded?

A big mistake, I fixed it yesterday.

>     * Why there is no `plugin` and `module`?
Apps plugin also does not have at the end of the line both values, I am stopping on update every due the same reason.

https://github.com/netdata/netdata/blob/ba350b7554d8a245bc488ed40409e594c4dd6f62/collectors/apps.plugin/apps_plugin.c#L3202

**Comment by thiagoftsm** [comment]
> Zero charts:
> Arch - 5.6.15
>     * file_open

I could not reproduct this problem when I tested on kernel 5.6.17.

>     * vfs_write_bytes
>     * vfs_read_bytes
>     * thread_create
>     * task_close (it works in general charts)

We really had a problem for `entry` mode with these charts, the fix that we brought with the release `0.4.3` fixed them.

> Ubuntu 18.04 - 4.15.0
> Ubuntu 20.04 - 5.4.0
> CentOS 7 - 3.10.0
> CentOS 8 - 4.18.0
> Debian 10 - 4.19.0
> Fedora 30 - 5.0.9
> OpenSUSE Leap 15.1 - 4.12.14
> 
>     * vfs_write_bytes
>     * vfs_read_bytes
>     * task_close (it works in general charts, except CentOS 8 and Debian 10)

The same thing happens on these distribution, they were affected for the same bug.

@vlvkobal , please, can you test them again? I am also testing on other kernels here.

**Comment by thiagoftsm** [comment]
> `apps.task_close` is still stuck at 0 in CentOS 8 but it works on ubuntu.
> 

CentOS 8 and Ubuntu have special kernels, I will take a look on it now.
For CentOS 8 we have a specific script to compile for it, I will take a look on it now.

**Comment by underhood** [comment]
@vlvkobal i am not sure we should test for Fed 30 as it is EOL soon

**Comment by underhood** [comment]
Fedora 32 AMD64 `apps.file_open` seems not right
Screenshot from 2020-06-11 15-22-00

**Comment by vlvkobal** [comment]
On Arch 5.6.15 `apps.file_deleted`, `apps.vfs_read_call`, `apps.vfs_read_bytes`, `apps.process_create`, `apps.task_close` are randomly accumulating some data. There was two CMake builds of Netdata:
image

**Comment by thiagoftsm** [comment]
> On Arch 5.6.15, Ubuntu 20.04 `apps.file_deleted`, `apps.vfs_read_call`, `apps.vfs_read_bytes`, `apps.process_create`, `apps.task_close` are randomly accumulating some data. There were two CMake builds of Netdata:
> image

The problem was fixed with the last commit.

**Pull request merged**